interfaces/prompting/constraints.go

```diff
@@ -0,0 +1,263 @@
+}
+
+// AbstractPermissionsFromAppArmorPermissions returns the list of permissions
+// corresponding to the given AppArmor permissions for the given interface.
+func AbstractPermissionsFromAppArmorPermissions(iface string, permissions interface{}) ([]string, error) {
+	return abstractPermissionsFromAppArmorFilePermissions(iface, permissions)
+}
+
+// abstractPermissionsFromAppArmorFilePermissions returns the list of permissions
+// corresponding to the given AppArmor file permissions for the given interface.
+func abstractPermissionsFromAppArmorFilePermissions(iface string, permissions interface{}) ([]string, error) {
+	filePerms, ok := permissions.(notify.FilePermission)
+	if !ok {
+		return nil, fmt.Errorf("failed to parse the given permissions as file permissions")
+	}
+	abstractPermsAvailable, exists := interfacePermissionsAvailable[iface]
+	if !exists {
+		// This should never happen, since iface is checked in the calling function.
+		return nil, fmt.Errorf("internal error: no permissions list defined for interface: %q", iface)
+	}
+	abstractPermsMap, exists := interfaceFilePermissionsMaps[iface]
+	if !exists {
+		// This should never happen, since iface is checked in the calling function.
+		return nil, fmt.Errorf("internal error: no file permissions map defined for interface: %q", iface)
+	}
+	if filePerms == notify.AA_MAY_OPEN {
+		// Should not occur, but if a request is received for only open, treat it as read.
+		filePerms = notify.AA_MAY_READ
+	}
+	// Discard Open permission; re-add it to the permission mask later
+	filePerms &= ^notify.AA_MAY_OPEN
+	abstractPerms := make([]string, 0, 1) // most requests should only include one permission
+	for _, abstractPerm := range abstractPermsAvailable {
+		aaPermMapping, exists := abstractPermsMap[abstractPerm]
+		if !exists {
+			// This should never happen, since permission mappings are
+			// predefined and should be checked for correctness.
+			return nil, fmt.Errorf("internal error: no permission map defined for abstract permission %q for interface %q", abstractPerm, iface)
+		}
+		if filePerms&aaPermMapping != 0 {
+			abstractPerms = append(abstractPerms, abstractPerm)
+			filePerms &= ^aaPermMapping
+		}
+	}
+	if filePerms != notify.FilePermission(0) {
+		return nil, fmt.Errorf("received unexpected permission for interface %q in AppArmor permission mask: %q", iface, filePerms)
+	}
+	if len(abstractPerms) == 0 {
+		origMask := permissions.(notify.FilePermission)
+		return nil, fmt.Errorf("no abstract permissions after parsing AppArmor permissions for interface: %q; original file permissions: %v", iface, origMask)
+	}
+	return abstractPerms, nil
+}
+
+// AbstractPermissionsFromList validates the given permissions list for the
+// given interface and returns a list containing those permissions in the order
+// in which they occur in the list of available permissions for that interface.
+func AbstractPermissionsFromList(iface string, permissions []string) ([]string, error) {
```

Comment:
True, the only use case for this right now is in `constraints.ValidateForInterface`, so this could be called `ValidatePermissions` as a method on the `Constraints` type.